webapp/src/snippetBuilder.tsx

```diff
@@ -0,0 +1,398 @@
+     */
+    componentDidMount() {
+        // Sets default values
+        this.buildDefaults();
+    }
+
+    /**
+     * @param output - Takes in a string and returns the tokenized output
+     * Loops over each token previously added to defaults and replaces with the answer value if one
+     * exists. Otherwise it replaces the token with the provided default value.
+     */
+    replaceTokens(tsOutput: string) {
+        const { answers, defaults } = this.state;
+        let tokenizedOutput = tsOutput;
+        const tokens = Object.keys(defaults);
+
+        // Replaces output tokens with answer if available or default value
+        for (let token of tokens) {
+            if (answers[token]) {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(answers[token]);
+            }
+            else {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(defaults[token]);
+            }
+        }
+
+        return tokenizedOutput;
+    }
+
+    /**
+     * 
+     * @param output - Accepts an output to convert to markdown
+     * This attaches three backticks to the front followed by an output type (blocks, lang)
+     * The current output is then tokenized and three backticks are appended to the end of the string.
+     */
+    generateOutputMarkdown(tsOutput: string) {
+        const { config } = this.state;
+        // Attaches starting and ending line based on output type
+        let md = `\`\`\`${config.outputType}\n`;
+        md += this.replaceTokens(tsOutput);
+        md += `\n\`\`\``;
+
+        return md
+    }
+
+    /**
+     * Hides the modal
+     */
+    hide() {
+        this.setState({
+            visible: false
+        });
+    }
+
+    /**
+     * 
+     * @param projectView - used to access what would traditionally be in the parent prop
+     * @param mainWorkspace  - used to append the final xml to the DOM
+     */
+    show(projectView: pxt.editor.IProjectView, mainWorkspace: Blockly.Workspace) {
+        pxt.tickEvent('snippetBuilder.show', null, { interactiveConsent: false });
+        this.setState({
+            visible: true,
+            mainWorkspace,
+            projectView,
+        });
+    }
+
+    /**
+     * Closes the modal
+     */
+    cancel() {
+        pxt.tickEvent("snippetBuilder.cancel", undefined, { interactiveConsent: true });
+        this.hide();
+    }
+
+    /**
+     * Takes the output from state, runs replace tokens, decompiles the resulting typescript
+     * and outputs the result as a Blockly xmlDOM. This then uses appendDomToWorkspace to attach 
+     * our xmlDOM to the mainWorkspaces passed to the component.
+     */
+    injectBlocksToWorkspace() {
+        const { mainWorkspace, tsOutput } = this.state;
+
+        compiler.getBlocksAsync()
+            .then(blocksInfo => compiler.decompileBlocksSnippetAsync(this.replaceTokens(tsOutput), blocksInfo))
+            .then(resp => {
```

Comment:
decompile can fail; needs to be handled gracefully.

Reply:
I added a catch so that decompile errors are handled.